src/main/java/com/epam/aidial/core/controller/DeploymentPostController.java

```diff
@@ -329,17 +329,29 @@ void handleProxyRequest(HttpClientRequest proxyRequest) {
      * Called when proxy received the response headers from the origin.
      */
     private void handleProxyResponse(HttpClientResponse proxyResponse) {
+        UpstreamRoute upstreamRoute = context.getUpstreamRoute();
+        Upstream currentUpstream = upstreamRoute.get();
         log.info("Received header from origin. Trace: {}. Span: {}. Key: {}. Deployment: {}. Endpoint: {}. Upstream: {}. Status: {}. Headers: {}",
                 context.getTraceId(), context.getSpanId(),
                 context.getProject(), context.getDeployment().getName(),
-                context.getDeployment().getEndpoint(), context.getUpstreamRoute().get().getEndpoint(),
+                context.getDeployment().getEndpoint(), currentUpstream == null ? "N/A" : currentUpstream.getEndpoint(),
                 proxyResponse.statusCode(), proxyResponse.headers().size());
 
-        if (context.getUpstreamRoute().hasNext() && isRetriableError(proxyResponse.statusCode())) {
+        int responseStatusCode = proxyResponse.statusCode();
+        if (isRetriableError(responseStatusCode)) {
+            upstreamRoute.fail(HttpStatus.fromStatusCode(responseStatusCode), UpstreamRoute.calculateRetryAfterSeconds(proxyResponse));
+            // get next upstream
+            upstreamRoute.next();
             sendRequest(); // try next
             return;
         }
 
+        if (responseStatusCode == 200) {
+            upstreamRoute.succeed();
+        } else if (isFailedStatusCode(responseStatusCode)) {
```

Comment:
So you want to remove the upstream from the list of available upstreams? Shouldn't we try to make one more try instead?

Reply:
if upstream failed with non-retryable status code we report used upstream anyway